### DISCUSSION TITLE
Create Github Action that uses a private Docker image

### DISCUSSION THREAD

**Discussion by helloimcarmine** [community/community#26399] [Actions]
<p>Hi all, I would like to create a Github action and to sell it on the marketplace. My action should be based on a Docker image that is stored on a private Docker registry and for licensing/security issues nobody could have direct access to it. Is there a way, also using Github secrets inside the action repository, to perform the access to such repository for example inside the action.yml file? My idea is to base the action on this Docker image and then to use a Python to script to perform the required operations calling a Python library installed into the container (all the business logic is inside the Docker image). Thanks in advance!</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/carminedamico">@carminedamico</a>,</p>
<p>About action, there are few points you need to understand:</p>
<ol>
<li>When using an action in the workflow, typically the following operations are included.</li>
</ol>
<ul>
<li>Check out the source code of the action to a special folder (<strong>_actions</strong>) under the runner’s working directory. This operation is done in the “<strong>Set up job</strong>” step.<br>
For example, when you run an action on the Linux runner,</li>
</ul>
<pre><code class="lang-yaml">- name: run action
  uses: my-actions/my-docker-action@master
  with:
    . . .
</code></pre>
<p>this action will be checked out to the path ‘<strong>/home/runner/work/_actions/my-actions/my-docker-action/master</strong>’.</p>
<ul>
<li>
<p>After checkout the source code, build the action if needed. Generally this is applied to build the container for the docker action, and the workflow run will generate a build step for each docker action.<br>
<img src="https://user-images.githubusercontent.com/46307996/181097482-959ad5e5-fa90-4dfd-b066-3bfdd694970e.png" alt="builds" data-base62-sha1="jU88H7KtSARXWT3V2As9CQBbhFy" width="475" height="71"></p>
</li>
<li>
<p>Execute the action with the inputs you set.</p>
</li>
</ul>
<ol start="2">
<li>
<p>If you store an action in a private repository, generally only this repository can use this action. If you want to use this private action in the workflows in other repositories, you need to use the <a href="https://github.com/actions/checkout#checkout-v2" rel="nofollow noopener"><strong>checkout</strong> action</a> with a personal access token (PAT) to checkout the source code of the action. The PAT should have the permissions (‘<strong>read</strong>’ at least) to access the private repository of the action.</p>
</li>
<li>
<blockquote>
<p><strong>Is there a way, also using Github secrets inside the action repository, to perform the access to such repository for example inside the action.yml file?</strong></p>
</blockquote>
</li>
</ol>
<p>No, we have no any way to do that. When using an action in a workflow, it is not directly executed in the action repository. The action is a step in the workflow, and every step in the workflow can only use the secrets in the repository where the workflow is running.</p>

**Comment by helloimcarmine** [comment]
<p>Hi <a class="mention" href="/brightran">@brightran</a> thank you for your reply! I don’t know if I got it wrong, but my problem is related to accessing a private Docker repository and not keeping the action repository private. The latter can be public.</p>
<p>What I’m interested in is access to use a Docker image stored in a private registry, without giving anyone who uses the action access to this registry. From your description, however, I understand that this is not possible, right?</p>

**Comment by BrightRan** [comment, marked as answer]
<p><a class="mention" href="/carminedamico">@carminedamico</a>,</p>
<p>Yes. Generally the <strong>base image</strong> used in a <strong>Dockerfile</strong> should be <strong>public</strong>.<br>
When using a Docker container action in the workflow, GitHub will do the following things during the workflow run:</p>
<ul>
<li>Pull the base image from the Docker registry.</li>
<li>Based on the base image, build the new image according to other steps and configurations set in Dockerfile.</li>
<li>Startup the container and execute the Docker container action.</li>
</ul>
<p>If you use a private image as the base image for the Docker container action, when pulling the base image, GitHub need to <strong>login</strong> the Docker registry at first authenticating with the <strong>user</strong> and <strong>password</strong> that can access the private Docker registry.</p>
<p>When you use this Docker container action in the workflow, you can try to add a step before the action step to execute the <strong>docker login</strong> command to login to the private Docker registry with the the <strong>user</strong> and <strong>password</strong>. Then execute the action to see if the authentication can work when pulling the base image.</p>
<p>There is a similar case as reference: <a href="https://stackoverflow.com/questions/57903470/how-build-a-dockerfile-using-base-image-from-private-registry" rel="nofollow noopener">https://stackoverflow.com/questions/57903470/how-build-a-dockerfile-using-base-image-from-private-registry</a></p>
<p>If you want other users can use this Docker container action, you need to share the authentication information (<strong>user</strong> and <strong>password</strong>) with the users, so that then can set the <strong>docker login</strong> step before the action step in their workflows.<br>
But this generally brings some obvious risks that a user who has gained the authentication may share the authentication to any other users that you do not grant to. The private image will become not “private”. Keeping it private no longer makes sense.</p>

**Reply by piotrekkr** [reply]
This answer is kind of misleading because when using `uses: docker://ghcr.io/repo/package:latest`, GitHub tries to load containers even before any step in job is executed. It fails because of this. What works:
1. login to ghcr using action
2. in next step run image `run: docker run IMAGE`

**Reply by divyangjp** [reply]
Only working solution atm until Github provides a solution.

**Reply by iainlane** [reply]
It also works to provide a shim `Dockerfile` which only has the `FROM` line referring to the private image.

That might be nicer in any event, since you can enable Dependabot for the shim Dockerfile and keep the image you're using updated that way. 🙂

**Reply by sheikhasim** [reply]
> It also works to provide a shim Dockerfile which only has the FROM line referring to the private image.

@iainlane Is this a working solution or a suggestion ? Facing unauthorized error when testing this against private registry.

**Reply by iainlane** [reply]
> > It also works to provide a shim Dockerfile which only has the FROM line referring to the private image.
> 
> 
> 
> @iainlane Is this a working solution or a suggestion ? Facing unauthorized error when testing this against private registry.

It works for me using a Docker action. To call the action you will need to auth to the registry prior to the step which `uses` the Docker action. Seems like the container used in the action is resolved early on if it's an image reference but if it's a Dockerfile it happens lazily.

**Reply by piotrekkr** [reply]
> > > It also works to provide a shim Dockerfile which only has the FROM line referring to the private image.
> > 
> > 
> > @iainlane Is this a working solution or a suggestion ? Facing unauthorized error when testing this against private registry.
> 
> It works for me using a Docker action. To call the action you will need to auth to the registry prior to the step which `uses` the Docker action. Seems like the container used in the action is resolved early on if it's an image reference but if it's a Dockerfile it happens lazily.

For me using `Dockerfile` with one line `FROM actual-image` does not provide any benefit because:
1. You still need to auth in step before.
1. GH need to actually build image first and then use it.
2. Actuall image used is "hidden" inside dockerfile.

Imho it is more clean to just run `docker run image` after auth step.

**Comment by helloimcarmine** [comment]
<p>Thank you very much!</p>

**Comment by leemeador** [comment]
<p><a class="mention" href="/brightran">@brightran</a> You spoke to the case where a Dockerfile is supplied in the github action. What I need to know is how it is done when you just want to use a docker image from a private registry and run shell commands inside it to do the work of the action. The example looks like this in the yaml file where you define the action:</p>
<pre><code class="lang-auto">runs: 
  using: 'docker'
  image: 'docker://myprivaterepo/someimage:notlatest'
</code></pre>
<p>Note that this became more important as of 2 Nov 2020 when Dockerhub placed pull limits on all nonauthenticated image pulls. In effect, everybody has to pay for an account and treat Dockerhub as a private registry when loading images to use in an action (as with every other case).</p>

**Reply by divyangjp** [reply]
@BrightRan what @leemeador asked in [this comment](https://github.com/orgs/community/discussions/26399#discussioncomment-3251766) is the problem which needs solution. Because the docker image from the action gets downloaded in the `Setup Job` step itself. Can not do auth to private docker registry before that.

**Reply by the-mentor** [reply]
did anyone of you (@divyangjp  @leemeador ) found a solution to this problem ?
